<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: **SUPER IMPORTANTE - notebooks sin nombre no serán revisados**

- Nombre de alumno 1: Angelo León Gianoni
- Nombre de alumno 2: Damián De Aguiar


### **Link de repositorio de GitHub:** [Repositorio](https://github.com/damiatus/labPrograCientifica)

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Jueves a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [ ]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
%pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de preparación y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [ ]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = '/content/drive/MyDrive/LabdeProgra/Lab3-Pandas'
except:
    print('Ignorando conexión drive-colab')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://i.imgur.com/44knnIf.png" width="500" height="500">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [ ]:
df_grades = pd.read_json(path + "/students_grades.json")
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
1         standard                    none        124           142   
2         standard                    none        156           196   
3         standard               completed         80            78   
4         standard                    none        131           166   
...            ...                     ...        ...           ...   
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
0               46  
1              142  
2              193  
3               81  
4              166  
...            ...  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76  

[1400 rows x 9 columns]


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **1000** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **1000** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier registro que posea un valor diferente a esta escala, **debería ser eliminado de la base de datos**.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [ ]:
def exploratory_data_analysis(dataframe):
    pd.options.display.max_rows = 200
    pd.set_option('display.max_columns', None)

    print('1.- El DataFrame tiene', dataframe.shape[0],'filas y', dataframe.shape[1], 'columnas\n')
    print('2.- El DataFrame está compuesto por las siguientes columnas: ', list(dataframe.columns),'\n')
    print('3.- Ejemplos de filas del DataFrame:\n')
    print('Primeras 5 filas:')
    display(dataframe.head(5))
    print('Últimas 5 filas:')
    display(dataframe.tail(5))
    print('Muestreo aleatorio de 5 filas:')
    display(dataframe.sample(5))
    print('4.- Descripción numérica del Dataframe:\n')
    display(dataframe.describe())
    print('5.- Cantidad de valores nulos por columna:')
    display(dataframe.isna().sum())
    print('6.- Cantidad de valores únicos por columna:')
    unique_values = dataframe.nunique()
    display(unique_values)
    print('7.- Filas totales duplicadas en el DataFrame:')
    display(dataframe.duplicated().sum())
    print('8.- Filas duplicadas en el DataFrame:')
    display(dataframe[dataframe.duplicated(keep=False)])

In [ ]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

In [ ]:
exploratory_data_analysis(df_grades)

1.- El DataFrame tiene 1400 filas y 9 columnas

2.- El DataFrame está compuesto por las siguientes columnas:  ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score'] 

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166

Últimas 5 filas:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76

Muestreo aleatorio de 5 filas:


names  gender race/ethnicity parental level of education  \
253   Tanya Favela  female        group D          associate's degree   
714   Charlie Luke    male        group E          associate's degree   
108   Debra Fannin  female        group B           bachelor's degree   
263   Connie Banda  female        group C                some college   
0    Rita Courtney  female        group B            some high school   

            lunch test preparation course math score reading score  \
253  free/reduced               completed         57            74   
714      standard               completed        131           124   
108      standard               completed         65            81   
263      standard               completed        122           166   
0        standard                    none         37            46   

    writing score  
253            76  
714           131  
108            81  
263           166  
0              46

4.- Descripción numérica del Dataframe:



names  gender race/ethnicity parental level of education  \
count             1400    1400           1400                        1400   
unique            1153       2              5                           6   
top     Robert Frazier  female        group C                some college   
freq                 3     723            452                         334   

           lunch test preparation course  math score  reading score  \
count       1400                    1400        1400           1400   
unique         2                       2          97             89   
top     standard                    none         193            193   
freq         904                     898         120            117   

        writing score  
count            1400  
unique             93  
top               193  
freq              133

5.- Cantidad de valores nulos por columna:


names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

6.- Cantidad de valores únicos por columna:


names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       97
reading score                    89
writing score                    93
dtype: int64

7.- Filas totales duplicadas en el DataFrame:


3

8.- Filas duplicadas en el DataFrame:


names gender race/ethnicity parental level of education  \
66    Anthony Malcomb   male        group C                some college   
562   Charles Jackson   male        group E            some high school   
816     Justin Harris   male        group D                some college   
827     Justin Harris   male        group D                some college   
900   Anthony Malcomb   male        group C                some college   
1223  Charles Jackson   male        group E            some high school   

             lunch test preparation course math score reading score  \
66    free/reduced                    none         35            28   
562       standard                    none         30            26   
816       standard                    none         40            42   
827       standard                    none         40            42   
900   free/reduced                    none         35            28   
1223      standard                    none         30            26   

     writing score  
66              27  
562             22  
816             38  
827             38  
900             27  
1223            22

**Justificación:**

```
9- Al revisar la funcion exploratory data analysis, se identifican 1400 filas, pero hay 1153 nombres distintos, lo que indica que hay nombres repetidos, eso sin considerar que el director menciono que hay 1000 matriculas, por lo que hay filas de mas. Tambien se puede apreciar que hay 3 filas duplicadas, ademas de que en cuanto a las notas, en describe se puede mostrar que en las 3 materias el maximo supera 100, lo que indica que esos datos fueron mal ingresados. No hay valores nulos en el dataframe.
10 - Con respecto al boxplox entregado por ChatGPT, se puede visualizar que los datos ingresados en las 3 materias estan incorrectos, debido a que los valores maximos superan el numero 100. Es mas, aproximadamente el tercer cuartil en cada materia esta alrededor del valor 100, lo que indica que un gran porcentaje esta mal puesto.
```

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y **cree una función** que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**
- [ ] Eliminar registros con valores nulos.
- [ ] Eliminar valores fuera del rango válido.
- [ ] Eliminar registros con nombres duplicados
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hints**

- Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.
- Cuidado con el orden en la limpieza de datos. ¿Que se debe eliminar primero, los nombres duplicados o los registros fuera del rango posible?


> Utilice la función creada en el punto 1 para verificar sus resultados.

**Respuesta:**

In [ ]:
def clean_data(dataframe):
    print("1- Eliminar valores nulos del dataframe")
    dataframe.replace("", np.nan, inplace=True)
    dataframe = dataframe.dropna()

    print("2- Eliminar valores fuera de rango")
    dataframe['math score'] = dataframe['math score'].astype(int)
    dataframe['reading score'] = dataframe['reading score'].astype(int)
    dataframe['writing score'] = dataframe['writing score'].astype(int)
    dataframe = dataframe[dataframe['math score'] <= 100]
    dataframe = dataframe[dataframe['reading score'] <= 100]
    dataframe = dataframe[dataframe['writing score'] <= 100]

    print("3- Eliminar registros con nombres duplicados")
    dataframe = dataframe.drop_duplicates(subset='names')

    return dataframe

In [ ]:
#Pruebe su código aquí
df_grades = clean_data(df_grades)
df_grades

1- Eliminar valores nulos del dataframe
2- Eliminar valores fuera de rango
3- Eliminar registros con nombres duplicados


<ipython-input-33-8660074584c3>:7: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-33-8660074584c3>:8: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-33-8660074584c3>:9: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none          37             46   
3         standard               completed          80             78   
5         standard                    none          73             66   
6         standard               completed          81             75   
7         standard               completed          94             86   
...            ...                     ...         ...            ...   
1392      standard                    none          69             75   
1394  free/reduced               completed          23             44   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1399      standard               completed          68             80   

      writing score  
0                46  
3                81  
5                63  
6                76  
7                87  
...             ...  
1392             71  
1394             36  
1396             81  
1397             62  
1399             76  

[1000 rows x 9 columns]

In [ ]:
exploratory_data_analysis(df_grades)

1.- El DataFrame tiene 1000 filas y 9 columnas

2.- El DataFrame está compuesto por las siguientes columnas:  ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score'] 

3.- Ejemplos de filas del DataFrame:

Primeras 5 filas:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none          37             46             46  
3  standard               completed          80             78             81  
5  standard                    none          73             66             63  
6  standard               completed          81             75             76  
7  standard               completed          94             86             87

Últimas 5 filas:


names  gender race/ethnicity parental level of education  \
1392   Richard Young    male        group D                 high school   
1394   Wanda Russell  female        group B                 high school   
1396  Marina Zeigler  female        group C           bachelor's degree   
1397   Laurie Carter  female        group B            some high school   
1399    Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1392      standard                    none          69             75   
1394  free/reduced               completed          23             44   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1399      standard               completed          68             80   

      writing score  
1392             71  
1394             36  
1396             81  
1397             62  
1399             76

Muestreo aleatorio de 5 filas:


names  gender race/ethnicity parental level of education  \
1337         Scott Price    male        group C                some college   
769         Laura Hunter  female        group E           bachelor's degree   
832          Eddie Strom    male        group E                some college   
1006  Richard Stackhouse    male        group A                some college   
876     Christopher Wall    male        group C                 high school   

             lunch test preparation course  math score  reading score  \
1337  free/reduced                    none          63             61   
769   free/reduced                    none          61             58   
832       standard                    none          69             60   
1006  free/reduced                    none          58             60   
876   free/reduced                    none          62             55   

      writing score  
1337             54  
769              62  
832              54  
1006             57  
876              55

4.- Descripción numérica del Dataframe:



math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

5.- Cantidad de valores nulos por columna:


names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

6.- Cantidad de valores únicos por columna:


names                          1000
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       81
reading score                    72
writing score                    77
dtype: int64

7.- Filas totales duplicadas en el DataFrame:


0

8.- Filas duplicadas en el DataFrame:


Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

**Justificación:**

```
Lo primero que se debe hacer es la eliminacion de registros fuera de rango, ya que estos podrian afectar a la eliminacion de filas duplicadas al poder ser registros mal ingresados en una sola linea. Por ello, se recomienda primero encargarse de los rangos en los valores, y luego eliminar duplicados.
```

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [ ]:
# Escriba su código aquí
def nota_chilena(dataframe_in, columna):
    dataframe_in['nota chilena'] = dataframe_in[columna] * 0.06 + 1 #transofmación a nota chilena
    dataframe_in['nota chilena'] = dataframe_in['nota chilena'].round(2) # redondeo a 2 decimales

    return dataframe_in


In [ ]:
#Muestre sus resultados aquí
nota_chilena(df_grades, 'math score')
nota_chilena(df_grades, 'reading score')
nota_chilena(df_grades, 'writing score')
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none          37             46   
3         standard               completed          80             78   
5         standard                    none          73             66   
6         standard               completed          81             75   
7         standard               completed          94             86   
...            ...                     ...         ...            ...   
1392      standard                    none          69             75   
1394  free/reduced               completed          23             44   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1399      standard               completed          68             80   

      writing score  nota chilena  
0                46          3.76  
3                81          5.86  
5                63          4.78  
6                76          5.56  
7                87          6.22  
...             ...           ...  
1392             71          5.26  
1394             36          3.16  
1396             81          5.86  
1397             62          4.72  
1399             76          5.56  

[1000 rows x 10 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [ ]:
#Escriba su código aquí
def alumnos_promedio(dataframe_in):
  dataframe_in['promedio'] = dataframe_in[['math score', 'reading score', 'writing score']].mean(axis=1)
  dataframe_in['promedio'] = dataframe_in['promedio'].round(2) # redondeo a 2 decimales
  return dataframe_in

In [ ]:
#Muestre sus resultados aquí
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none          37             46             46   
3  standard               completed          80             78             81   
5  standard                    none          73             66             63   
6  standard               completed          81             75             76   
7  standard               completed          94             86             87   

   nota chilena  promedio  
0          3.76     43.00  
3          5.86     79.67  
5          4.78     67.33  
6          5.56     77.33  
7          6.22     89.00

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [ ]:
#Escriba su código aquí

# Muestre sus resultados aquí


def reprobados_destacados(dataframe_in):
    dataframe1 = dataframe_in[dataframe_in['promedio'] < 40] # primer df

    df_grades2 = dataframe_in.copy() #copiamos el df original para modificarlo
    df_grades2[['promedio']] = df_grades2[['promedio']].map(lambda x: 'R' if x < 40 else x) #aplicamos un mapeo reemplazando los valores bajo 4 por R
    dataframe2 = df_grades2 # segundo df

    dataframe3 = dataframe_in[dataframe_in['promedio'] > 5.8] # tercer df
    dataframe3 = dataframe3.sort_values(by='promedio', ascending=False)
    return {'DataFrame 1': dataframe1,'DataFrame 2': dataframe2, 'DataFrame 3': dataframe3}


In [ ]:
#mostrar resulados
reprobados_destacados(df_grades)

{'DataFrame 1':                   names  gender race/ethnicity parental level of education  \
 55      Rodger Talavera    male        group B                 high school   
 66      Anthony Malcomb    male        group C                some college   
 196     Patricia Briney  female        group B                some college   
 311      Gerald Brunner    male        group D                 high school   
 312   Barbara Henriques  female        group C                some college   
 357      Paula Barriere  female        group D           bachelor's degree   
 358        Marina Smith  female        group D          associate's degree   
 372         John Gunter    male        group A                some college   
 412    Flossie Streeter  female        group C            some high school   
 558    Johnathan Bishop    male        group D           bachelor's degree   
 562     Charles Jackson    male        group E            some high school   
 679    Luetta Caponigro  female     

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [ ]:
# Escriba su código para quintiles aquí
df_grades['GPA cuantiles'] = pd.qcut(df_grades['promedio'], q=5, labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])


In [ ]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="promedio"),
    x="promedio",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

```
Escriba aqui su respuesta


Se aprecia como entre se genera una tendencia a peores notas entre los hijos de padres con menores nivel educacional. Se aprecia como los hijos de estos padres tienden a tener peores notas.
Por otro lado, también se observa que esto también se cumple pero en menor medida para los alumnos de mejores notas, siendo estos menos cantidad en los casos en que los padres tienen menos estudios.

```

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

**Respuesta:**

In [92]:
# Inserte su respuesta aquí
df_grades_1= pd.read_json(path + "/students_grades_1.json")
df_grades_2= pd.read_json(path + "/students_grades_2.json")

df_grades= pd.concat([df_grades_1, df_grades_2], axis=0, ignore_index=True)
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
995      standard                    none        5.14           5.50   
996  free/reduced               completed        2.38           3.64   
997  free/reduced               completed        4.96           5.44   
998      standard               completed        4.24           4.66   
999      standard               completed        5.08           5.80   

     writing score   GPA GPA cuantiles  
0             3.76  3.58      Bajísimo  
1             5.86  5.78          Alto  
2             4.78  5.04         Medio  
3             5.56  5.64          Alto  
4             6.22  6.34      Muy alto  
..             ...   ...           ...  
995           5.26  5.30         Medio  
996           3.16  3.06      Bajísimo  
997           5.86  5.42          Alto  
998           4.72  4.54          Bajo  
999           5.56  5.48          Alto  

[1000 rows x 11 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos]**


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


**Respuesta:**

In [93]:
mapeo = {
    'some high school': 'school',
    'some college': 'school',
    'high school': 'school',
    'bachelor\'s degree': 'college',
    'associate\'s degree': 'college',
    'master\'s degree': 'postgraduate'
}
df_grades['simple parental level of education'] = df_grades['parental level of education'].map(mapeo)

promedio_notas = df_grades.groupby(['race/ethnicity', 'simple parental level of education'])[['math score', 'reading score', 'writing score']].mean().round(2)

numero_alumnos = df_grades.groupby(['race/ethnicity', 'simple parental level of education']).size()

total_alumnos = df_grades.shape[0]

percentage = (numero_alumnos / total_alumnos * 100).map(lambda x: f'{x:.2f}%')
numero_alumnos_df = pd.DataFrame({
    'count': numero_alumnos,
    'percentage': percentage
})

tabla = pd.merge(promedio_notas, numero_alumnos_df, on=['race/ethnicity', 'simple parental level of education'])

tabla = tabla.sort_values(by=['race/ethnicity', 'simple parental level of education'])

In [94]:
tabla

math score  reading score  \
race/ethnicity simple parental level of education                              
group A        college                                   4.83           5.05   
               postgraduate                              4.46           4.88   
               school                                    4.65           4.81   
group B        college                                   5.03           5.24   
               postgraduate                              5.03           5.81   
               school                                    4.69           4.90   
group C        college                                   5.03           5.36   
               postgraduate                              5.02           5.23   
               school                                    4.74           5.00   
group D        college                                   5.06           5.22   
               postgraduate                              5.35           5.63   
               school                                    4.99           5.13   
group E        college                                   5.53           5.45   
               postgraduate                              5.48           5.93   
               school                                    5.35           5.27   

                                                   writing score  count  \
race/ethnicity simple parental level of education                         
group A        college                                      4.95     26   
               postgraduate                                 5.06      3   
               school                                       4.66     60   
group B        college                                      5.16     61   
               postgraduate                                 5.63      6   
               school                                       4.79    123   
group C        college                                      5.33    118   
               postgraduate                                 5.17     19   
               school                                       4.89    182   
group D        college                                      5.24     78   
               postgraduate                                 5.78     23   
               school                                       5.11    161   
group E        college                                      5.43     57   
               postgraduate                                 5.83      8   
               school                                       5.11     75   

                                                  percentage  
race/ethnicity simple parental level of education             
group A        college                                 2.60%  
               postgraduate                            0.30%  
               school                                  6.00%  
group B        college                                 6.10%  
               postgraduate                            0.60%  
               school                                 12.30%  
group C        college                                11.80%  
               postgraduate                            1.90%  
               school                                 18.20%  
group D        college                                 7.80%  
               postgraduate                            2.30%  
               school                                 16.10%  
group E        college                                 5.70%  
               postgraduate                            0.80%  
               school                                  7.50%

In [95]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

Observamos que el grupo B mejora su promedio de notas a medida que el nivel educativo de los padres aumenta. Esta mejora comienza cuando los padres tienen educación secundaria o universitaria, y se vuelve aún más significativa cuando los padres tienen un doctorado. En ese caso, el grupo B pasa de ser el segundo con peores notas a ser el segundo con mejores notas.

Además, el desempeño promedio en educación secundaria y universitaria sigue un patrón similar entre etnias, en el orden creciente de A, B, C, D y E. Sin embargo, en el caso de los padres con doctorado, el orden cambia un poco, con los grupos B y D intercambiando posiciones. El grupo B muestra una notable mejora en comparación con sus dos niveles educativos anteriores.

Por otro lado, en la etnia A, las notas medianas se mantienen constantes sin importar el nivel educativo de los padres. En la etnia B, como ya mencionamos, las notas aumentan con el nivel educativo de los padres. En la etnia C, resulta curioso ver una disminución en la media de notas cuando los padres tienen un doctorado. Finalmente, los grupos D y E muestran medias similares cuando los padres tienen educación secundaria o universitaria, pero ambos se benefician notablemente si los padres tienen un doctorado, ya que su promedio de notas aumenta.

**Resultados Esperados**

Esta tabla ilustra la estructura del dataframe a conseguir. Los valores son referenciales, no necesariamente son los esperados.

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.83 |            5.05 |            4.95 |  4.94 |      26 | 2.6 %        |
|  1 |                  | postgraduate                         |         4.46 |            4.88 |            5.06 |  4.80 |       3 | 0.3 %        |
|  2 |                  | school                               |         4.65 |            4.81 |            4.66 |  4.71 |      60 | 6.0 %        |
|  3 | group B          | college                              |         5.03 |            5.24 |            5.16 |  5.14 |      61 | 6.1 %        |
|  4 |                  | postgraduate                         |         5.03 |            5.81 |            5.63 |  5.49 |       6 | 0.6 %        |
|  5 |                  | school                               |         4.69 |            4.90 |            4.79 |  4.79 |     123 | 12.3 %       |
|  6 | group C          | college                              |         5.03 |            5.36 |            5.33 |  5.24 |     118 | 11.8 %       |
|  7 |                  | postgraduate                         |         5.02 |            5.23 |            5.17 |  5.14 |      19 | 1.9 %        |
|  8 |                  | school                               |         4.74 |            5.00 |            4.89 |  4.88 |     182 | 18.2 %       |
|  9 | group D          | college                              |         5.06 |            5.22 |            5.24 |  5.17 |      78 | 7.8 %        |
| 10 |                  | postgraduate                         |         5.35 |            5.63 |            5.78 |  5.59 |      23 | 2.3 %        |
| 11 |                  | school                               |         4.99 |            5.13 |            5.11 |  5.08 |     161 | 16.1 %       |
| 12 | group E          | college                              |         5.53 |            5.45 |            5.43 |  5.47 |      57 | 5.7 %        |
| 13 |                  | postgraduate                         |         5.48 |            5.93 |            5.83 |  5.74 |       8 | 0.8 %        |
| 14 |                  | school                               |         5.35 |            5.27 |            5.11 |  5.25 |      75 | 7.5 %        |


### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene registros de alumnos que ya no forman parte de la universidad, por lo que no deben ser parte de este análisis.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados o **alumnos que no pertenecen a la universidad**).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

**Respuesta:**

In [ ]:
# Inserte su respuesta aquí

#df_grades = pd.read_csv(path + "/students_grades.csv")
other_grades = pd.read_csv(path + '/other_grades.csv')

df_grades_outer = pd.merge(df_grades, other_grades, on='names', how='outer')
df_grades_left = pd.merge(df_grades, other_grades, on='names', how='left')
df_grades_right = pd.merge(df_grades, other_grades, on='names', how='right')
df_grades_inner = pd.merge(df_grades, other_grades, on='names', how='inner')

In [ ]:
df_grades_outer

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1145     Carson Brooks     NaN            NaN                         NaN   
1146         Ruby Reed     NaN            NaN                         NaN   
1147    Zachary Torres     NaN            NaN                         NaN   
1148       Eliana Gray     NaN            NaN                         NaN   
1149       Gavin James     NaN            NaN                         NaN   

         lunch test preparation course  math score  reading score  \
0     standard                    none        37.0           46.0   
1     standard               completed        80.0           78.0   
2     standard                    none        73.0           66.0   
3     standard               completed        81.0           75.0   
4     standard               completed        94.0           86.0   
...        ...                     ...         ...            ...   
1145       NaN                     NaN         NaN            NaN   
1146       NaN                     NaN         NaN            NaN   
1147       NaN                     NaN         NaN            NaN   
1148       NaN                     NaN         NaN            NaN   
1149       NaN                     NaN         NaN            NaN   

      writing score  nota chilena  promedio GPA cuantiles  science score  \
0              46.0          3.76     43.00      Bajísimo           2.50   
1              81.0          5.86     79.67          Alto           4.36   
2              63.0          4.78     67.33         Medio           4.48   
3              76.0          5.56     77.33          Alto           5.32   
4              87.0          6.22     89.00      Muy alto           6.64   
...             ...           ...       ...           ...            ...   
1145            NaN           NaN       NaN           NaN           4.72   
1146            NaN           NaN       NaN           NaN           1.71   
1147            NaN           NaN       NaN           NaN           4.48   
1148            NaN           NaN       NaN           NaN           6.91   
1149            NaN           NaN       NaN           NaN           6.60   

      history score  
0              3.80  
1              5.20  
2              3.80  
3              5.00  
4              4.60  
...             ...  
1145           1.54  
1146           4.59  
1147           6.92  
1148           5.41  
1149           6.06  

[1150 rows x 14 columns]

In [ ]:
df_grades_left

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none          37             46   
1        standard               completed          80             78   
2        standard                    none          73             66   
3        standard               completed          81             75   
4        standard               completed          94             86   
..            ...                     ...         ...            ...   
995      standard                    none          69             75   
996  free/reduced               completed          23             44   
997  free/reduced               completed          66             74   
998      standard               completed          54             61   
999      standard               completed          68             80   

     writing score  nota chilena  promedio GPA cuantiles  science score  \
0               46          3.76     43.00      Bajísimo           2.50   
1               81          5.86     79.67          Alto           4.36   
2               63          4.78     67.33         Medio           4.48   
3               76          5.56     77.33          Alto           5.32   
4               87          6.22     89.00      Muy alto           6.64   
..             ...           ...       ...           ...            ...   
995             71          5.26     71.67         Medio           4.72   
996             36          3.16     34.33      Bajísimo           1.96   
997             81          5.86     73.67          Alto           4.54   
998             62          4.72     59.00          Bajo           3.58   
999             76          5.56     74.67          Alto           4.66   

     history score  
0              3.8  
1              5.2  
2              3.8  
3              5.0  
4              4.6  
..             ...  
995            4.2  
996            3.6  
997            4.1  
998            4.3  
999            5.3  

[1000 rows x 14 columns]

In [ ]:
df_grades_right

names  gender race/ethnicity parental level of education  \
0           Pam Otoole  female        group B           bachelor's degree   
1          Diane Olson  female        group C                some college   
2       Sarah Flannery  female        group B             master's degree   
3     Stanley Chappell    male        group A          associate's degree   
4        Arnold Dodson    male        group C                some college   
...                ...     ...            ...                         ...   
1145     Carson Brooks     NaN            NaN                         NaN   
1146         Ruby Reed     NaN            NaN                         NaN   
1147    Zachary Torres     NaN            NaN                         NaN   
1148       Eliana Gray     NaN            NaN                         NaN   
1149       Gavin James     NaN            NaN                         NaN   

             lunch test preparation course  math score  reading score  \
0         standard                    none        72.0           72.0   
1         standard               completed        69.0           90.0   
2         standard                    none        90.0           95.0   
3     free/reduced                    none        47.0           57.0   
4         standard                    none        76.0           78.0   
...            ...                     ...         ...            ...   
1145           NaN                     NaN         NaN            NaN   
1146           NaN                     NaN         NaN            NaN   
1147           NaN                     NaN         NaN            NaN   
1148           NaN                     NaN         NaN            NaN   
1149           NaN                     NaN         NaN            NaN   

      writing score  nota chilena  promedio GPA cuantiles  science score  \
0              74.0          5.44     72.67          Alto           4.42   
1              88.0          6.28     82.33      Muy alto           3.88   
2              93.0          6.58     92.67      Muy alto           5.32   
3              44.0          3.64     49.33      Bajísimo           3.82   
4              75.0          5.50     76.33          Alto           4.60   
...             ...           ...       ...           ...            ...   
1145            NaN           NaN       NaN           NaN           4.72   
1146            NaN           NaN       NaN           NaN           1.71   
1147            NaN           NaN       NaN           NaN           4.48   
1148            NaN           NaN       NaN           NaN           6.91   
1149            NaN           NaN       NaN           NaN           6.60   

      history score  
0              4.00  
1              4.80  
2              6.10  
3              3.70  
4              5.20  
...             ...  
1145           1.54  
1146           4.59  
1147           6.92  
1148           5.41  
1149           6.06  

[1150 rows x 14 columns]

In [ ]:
df_grades_inner

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none          37             46   
1        standard               completed          80             78   
2        standard                    none          73             66   
3        standard               completed          81             75   
4        standard               completed          94             86   
..            ...                     ...         ...            ...   
995      standard                    none          69             75   
996  free/reduced               completed          23             44   
997  free/reduced               completed          66             74   
998      standard               completed          54             61   
999      standard               completed          68             80   

     writing score  nota chilena  promedio GPA cuantiles  science score  \
0               46          3.76     43.00      Bajísimo           2.50   
1               81          5.86     79.67          Alto           4.36   
2               63          4.78     67.33         Medio           4.48   
3               76          5.56     77.33          Alto           5.32   
4               87          6.22     89.00      Muy alto           6.64   
..             ...           ...       ...           ...            ...   
995             71          5.26     71.67         Medio           4.72   
996             36          3.16     34.33      Bajísimo           1.96   
997             81          5.86     73.67          Alto           4.54   
998             62          4.72     59.00          Bajo           3.58   
999             76          5.56     74.67          Alto           4.66   

     history score  
0              3.8  
1              5.2  
2              3.8  
3              5.0  
4              4.6  
..             ...  
995            4.2  
996            3.6  
997            4.1  
998            4.3  
999            5.3  

[1000 rows x 14 columns]

        Explique sus resultados acá

1. El outer join mantiene todas las filas de ambos dataframes. Si un alumno está en uno pero no en el otro, la información del datafreame en el que no se encuentra el alumno será NaN.

2. El left join mantiene todas las filas del dataframe de la izquierda (df_grades) y agrega la información del dataframe de la derecha (other_grades) cuando hay coincidencia.

3. El right join hace el opuesto del left join: mantiene todas las filas del dataframe de la derecha (other_grades) y agrega la información del dataframe de la izquierda (df_grades) donde hay coincidencia.

4. El inner join solo mantiene las filas donde hay coincidencia en ambos dataframes.


Para este caso, vemos como tanto el left join como el inner join dan resultado a la misma cantidad de filas. En este caso particular se produce ya que las filas de df_grades son un subonjunto de other_grades. Igualmente, por riesgos de que el nuevo dataframe other_grades no contenga datos que sí están en df_grades, lo mpas recomendable es utilizar left join dado que el dataframe que sabemos está limpio y tiene los datos correctos es df_grades.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>